libp2p/protocols/rendezvous.nim

```diff
@@ -312,39 +127,39 @@ proc checkPeerRecord(spr: seq[byte], peerId: PeerId): Result[void, string] =
 
 proc sendRegisterResponse(conn: Connection,
                           ttl: uint64) {.async.} =
-  let msg = encode(Message(
-        msgType: MessageType.RegisterResponse,
-        registerResponse: some(RegisterResponse(status: Ok, ttl: some(ttl)))))
-  await conn.writeLp(msg.buffer)
+  let msg = Protobuf.encode(Message(
+        msgType: RdvRegisterResponse,
+        registerResponse: RegisterResponse(status: ResponseOk, ttl: ttl)))
+  await conn.writeLp(msg)
 
 proc sendRegisterResponseError(conn: Connection,
-                               status: ResponseStatus,
+                               status: uint64,
                                text: string = "") {.async.} =
-  let msg = encode(Message(
-        msgType: MessageType.RegisterResponse,
-        registerResponse: some(RegisterResponse(status: status, text: some(text)))))
-  await conn.writeLp(msg.buffer)
+  let msg = Protobuf.encode(Message(
```

Comment:
cc @Menduist this is where you'd use a streaming writer instead - ie each of these functions is pointlessly inefficient

Reply:
So you would imagine a `Protobuf.streamEncode` macro that takes
```nim
Protobuf.streamEncode(Message(
  msgType: RdvRegisterResponse,
  registerResponse: RegisterResponse(status: status, text: text))
)
# and transforms into
block:
  var pb = initProtoBuf()
  pb.writeField(1, RdvRegisterResponse)
  pb.writeField(2, Protobuf.streamEncode(RegisterResponse(status: status, text: text))
  pb.finish()
  pb.buffer
```
? That shouldn't be that hard to do

Reply:
Well, or some variations on that theme:
```nim
Protobuf.encode(Message, encoder):
  encoder.write(fieldname1, value)
  encoder.write(somesubobject, subencoder):
    subencoder.write(subfield, value)
```

```nim
Protobuf.encode(Message, encoder)
    .field1(value).
    .field2(subobject):
       suboject
          .subField(value)
    .field3(value)
```

not sure, haven't really looked into it, though the obvious point is to grab the field number and encoding from the object definition and for that you need the type and the field name - the rest is really a syntax game.

There are some caveats though - streaming inherently becomes more "freeform" in that you can repeat the same field multiple times (this is valid in pb even for non-repeated) and there are some quirks with packed fields (which are length-prefixed and therefore would need an intermediate store - either the fancy but buggy faststreams buffers or something equivalent.

Reply:
When I look at your example, maybe transforming the object constructor isn't that bad :) That said, similar caveats apply: for example, to correctly encode a nested object you need to evaluate it twice: once to know if it has any non-default values and the second time to actually write the bytes - do you then evaluate the "source" twice or introduce a temporary buffer? tradeoffs, tradeoffs...